Offline sync via NSPersistentHistory, support for CloudKit Sharing

deeje opened the pull request:
I really liked your project, and have spent some time adding to it.

- refactored names to standardize on "push" and "pull" to help me better understand the flow of data
- refactored isPublic to support pushing to private, public, or both databases
- incorporated support for relationships via Oleg Müller
- added support for NSPersistentHistoryTracking, to enable offline sync
- added support for CloudKit Sharing via the sharedDatabase
- updated unit tests
- updated the example app

I still need to update the following
- online tests 
- documentation
- pass Travis CI
but wanted to get this PR up to gather thoughts.  Lemme know.

Scytalion committed:
supporting transformable attributes

Scytalion committed:
adding fixes for fixing relationships

# Conflicts:
#	Source/Classes/Pull/SubOperations/RecordToCoreDataOperation.swift

Scytalion committed:
error handling

Commit:
normalize on Push & Pull nomenclature

Commit:
ignore .DS_Store files

Commit:
more nomenclature changes

Commit:
still more nomenclature changes

Commit:
don’t need to use self.

Commit:
enable persistent history in Core Data

Commit:
move the push code outside of the didSave handler

deeje committed:
update to Swift 4.2

deeje committed:
update CloudCore to Swift 4.2

deeje committed:
support for push to CloudKit via PersistentHistory

deeje committed:
CloudCore.usePersistenHistoryForPush

deeje committed:
fix delete

deeje committed:
processPersistentHistory when reachability changes

deeje committed:
catch overall CKModify record operation errors

deeje committed:
delete CK records from tombstone recordData

deeje committed:
don’t push all if using persistent history

and clean up private MOCs

deeje committed:
rename to PullResult for consistency

deeje committed:
more normalizing on push/pull terminology

deeje committed:
TODO: optimize by using change.updatedProperties

deeje committed:
actually delete history as we push

deeje committed:
use recordNames to match CK and CD objects

deeje committed:
add ServiceAttributeNames.contains(…)

deeje committed:
use existing recordName if it exists

deeje committed:
support for public push

deeje committed:
replace .isPublic with ServiceAttributeNames.scopes

allow Core Data schema to define which scopes to push

deeje committed:
only setRecordInfo for scopes

deeje committed:
use container.databse(with: scope)

deeje committed:
with recordName, we don’t need to store recordIDs

deeje committed:
remove unused CKRecordID extensions

deeje committed:
add ServiceAttributeNames.ownerName

deeje committed:
properly establish parent-child relationships

deeje committed:
making some extension funcs public, for sharing

deeje committed:
support for (shared) database changes

deeje committed:
implement recordZoneWithIDWasDeletedBlock

other user stopped sharing something, so clear it out locally

deeje committed:
more resilience when error during fetchHistory

deeje committed:
update test cases to newer APIs

deeje committed:
update Example to latest CloudCore

deeje committed:
update ReadMe to latest CloudCore

deeje committed:
reset Example bundle ID

deeje committed:
update author list

deeje committed:
[deeje] updated ReadMe to properly reflect status

Sorix commented:
Thank you a lot for such big PR, I have not a lot of time for that project right now 🙁, so I can't deeply test everything you proposed. I will check if your code is working on a next week, it will outdate all other PRs also.

Do you want a second maintainer status? I can give a push permission for you after code review, so you can push / accept other Pull Requests.

deeje committed:
optimize by using change.updatedProperties

deeje committed:
update CloudKitTests

deeje committed:
more updates to CloudKitTests

deeje committed:
updating swift-version and travis.yml to latest

deeje committed:
tell travis to build the CloudCore.xcworkspace

deeje committed:
don’t point travis to “Example/Podfile”

deeje committed:
update test code to Swift 4.2

deeje committed:
enable NSPersistentHistoryTrackingKey in iOS 11

deeje committed:
disabling build Example in travis

deeje committed:
Only use Reachability on iOS

deeje committed:
remove Reachability as a dependency

its up to the host app to change CloudCore.isOnline as appropriate for history-driven push

deeje committed:
update travis to build project instead of workspace

deeje committed:
clean out the last of the Cocoapods from project

deeje committed:
fix use of NSPersistentHistory in CoreDataTestCase

deeje committed:
updating podspec for Travis CI

deeje committed:
skip unkown record values, probably newer schema

Sorix commented:
@deeje, please look at https://github.com/Sorix/CloudCore/pull/51#issuecomment-445068236

deeje committed:
Properly ignore unknown data from CloudKit

deeje committed:
enable editing of shared data

deeje committed:
Tokens now fully managed by CloudCore

deeje committed:
remove applicationWillTerminate from Example.app

deeje committed:
Pull records serially, RecordToCD.performAndWait

deeje committed:
on launch, check for pending changes

hyerra committed:
Make sure only app extension APIs only are used to prevent warnings being thrown in WatchKit.

deeje committed:
support for query subscriptions on public database

deeje committed:
(silence a warning)

deeje committed:
don’t bother updating deleted records

deeje committed:
add a matching unsubscribe call, generate same id

deeje committed:
remove unused/duplicate typealias

deeje committed:
making RecordToCoreDataOperation for app use

deeje committed:
udpated ReadMe to include discussion of iOS 13

deeje committed:
added additional note regarding iOS 13 native sync

deeje committed:
update podspec to use Swift 5.0

deeje committed:
update Example project to Swift 5

deeje committed:
remove .swift-version, use version in podspec

deeje committed:
update to Swift 5.0

deeje committed:
update tests to Swift 5.0

deeje committed:
Merge tag '3.0' into develop

3.0

deeje committed:
Merge branch 'release/3.0'

deeje committed:
use Xcode 10.3 for Swift 5.0

deeje committed:
Release 3.0.1

deeje committed:
update ReadMe CocoaPods instructions

deeje committed:
Merge branch 'master' into develop

deeje committed:
update to latest KeyedArchiver APIs

deeje committed:
refactor to use ZoneConfiguration, not ZoneOptions

deeje committed:
refactor to use CKOperation.Configuration

deeje committed:
update the license

deeje committed:
require iOS 13, macOS 10.15, watchOS 6, Swift 5.1

deeje committed:
enable watchOS 6 support

deeje committed:
gracefully handle unknown references

deeje committed:
fix broken tokens

deeje committed:
fix bug in CoreData where _defaultOwner != __defaultOwner__

deeje committed:
fix a race condition in processChanges ?

deeje committed:
make CloudCoreConfig fully public

deeje committed:
implement PullResult.wkBackgroundFetchResult

deeje committed:
don’t assume CKContainer.default(), use Config

deeje committed:
container and db op queues, QoS = .userInteractive

this commit improves overall performance of sync in CloudCore, and makes sync of a large iCloud account (like mine) work on watchOS.

deeje committed:
ability to pull a root record, useful for sharing

deeje committed:
formally define persistentHistoryTokenKey

deeje committed:
save backghroundContexts in their own thread

deeje committed:
ensure all Ops ask for background task time

deeje committed:
NSPersistentContainer.performBackgroundPushTask

convenience func to automatically name the background context, which triggers the push to CloudKit

deeje committed:
grab ownerName, targetScope inside context!

deeje committed:
eliminate warning re CloudCoreDelegate: AnyObject

deeje committed:
update Example to use Connectivity cocoapod

deeje committed:
update Xcode project, target iOS 13, watchOS 6

deeje committed:
use Xcode 12.5 for travis

deeje committed:
CloudCore now fetches userRecordID

deeje committed:
add Sharing support

deeje committed:
update Example with Sharing support

deeje committed:
updated ReadMe to include info about Sharing

deeje committed:
document windowScene(userDidAcceptCloudKitShare)

if you’re using scenes, implement this, otherwise, implement the app delegate version

deeje committed:
make Tokens atomic

deeje committed:
dont hide potential bugs behind try?

deeje committed:
don’t start uploading until zone is created

deeje committed:
(remove unnecessary () from async call site )

deeje committed:
add PullChangesOp directly to queue

deeje committed:
save context on completing recordZoneChangesOp

deeje committed:
eliminate ErrorBlocvkProxy

deeje committed:
use the context in processMissingReferences

deeje committed:
add handlers for when share is saved or stopped

deeje committed:
DRY when fetching the share

deeje committed:
don’t try to upate object when sharing stops

deeje committed:
update example when displaying sharing UI

deeje committed:
DRY, call shareDatabaseAndRecordID instead

deeje committed:
call setShareRecore on start/stop sharing

deeje committed:
UIViewController. iCloudAvailable(withPrompt…)

sometimes you only want to show an alert once per launch or once per install or not at all

deeje committed:
when the zone is first created, do an initial pull

this also has the effect of calling the delegate as expected

deeje committed:
(remove an duplicate line of code)

deeje committed:
backgroundPushTask is always merge PropertyObjectTrump

deeje committed:
all pull operations should be serialized

in case we get called to sync while we're still syncing, eliminates potential duplicates

deeje committed:
bumping podspec version to 4.0

deeje committed:
update license & readme

deeje committed:
Merge pull request #22 from deeje/feature/4.0

Feature/4.0

deeje committed:
set state of AsyncOp in its own thread

NOT in its parentContext

deeje committed:
still trying to resolve rando crash re missing refs

deeje committed:
align podspec targets with project targets

deeje committed:
Merge pull request #23 from deeje/feature/4.0.1

4.0.1 - bug fixes

deeje committed:
fix for CloudKit error 3, not online when using cellular

qualityOfService = .userInitiated

via https://stackoverflow.com/questions/34189458/cloudkit-ckdatabaseoperation-not-working-on-cellular

deeje committed:
more qualityOfService = .userInitiated

deeje committed:
CloudCore.perform { configuredcontainer in }

deeje committed:
fix use of #availability

deeje committed:
explicitly import CloudKit in Example AppDelegate

deeje committed:
bump podspec s.version = 4.0.2

deeje committed:
Merge pull request #2 from hyerra/master

Fix App Extension API only warning in WatchKit

deeje committed:
Merge remote-tracking branch 'deeje/master' into feature/4.0.2

deeje committed:
Merge pull request #24 from deeje/feature/4.0.2

Fix for CloudKit error 3 when using cellular

deeje committed:
read from record.encryptedValues

note that record.allKeys() includes record.encryptedValues.allKeys()

deeje committed:
if allowsCloudEncryption, store in encryptedValues

deeje committed:
update Example app to support encrypted fields

deeje committed:
(update sharing type to match bundleID)

deeje committed:
handle CKErrorUserDidResetEncryptedData

deeje committed:
another URL re CKErrorUserDidResetEncryptedDataKey

deeje committed:
update project

deeje committed:
Update ReadMe to include 'Allows Cloud Encryption’

deeje committed:
update podspec version to 4.1

deeje committed:
Merge pull request #27 from deeje/feature/EncryptedFields

EncryptedValues

deeje committed:
proper #available for encryption support

deeje committed:
clean up deprecated CKRecord_Reference_Action

deeje committed:
update podspec version to 4.1.1

deeje committed:
masked attributes

use desiredKeys in fetch, skip maskedUpload in modify

deeje committed:
update version to 5.0.0

deeje committed:
updating ReadMe with Scope, Maskable Attributes

deeje committed:
(add missing classes, update names to project)

deeje committed:
CloudCoreCacheable, manual upload & download

deeje committed:
trigger upload after successful push

deeje committed:
show progress during upload, automatic download

deeje committed:
no refresh needed when triggering cacheable.upload

deeje committed:
remove data file when cacheable is deleted

deeje committed:
(we’ll come back to .unload)

deeje committed:
when caching fails, save lastErrorMessage

deeje committed:
retry on rateLimit or zoneBusy errors

deeje committed:
(call UI completion handlers on main thread)

deeje committed:
merged can’t fail during pull

deeje committed:
store cacheables in App Support directory

deeje committed:
update ReadMe with Cacheable Assets info

deeje committed:
MUST override awakeFromInsert & prepareForDeletion

deeje committed:
very important to preserveAfterDeletion

deeje committed:
move default values into awakeFromInsert

deeje committed:
nope, developer must manually set cache state

deeje committed:
use FileManager to find, and create, user paths

deeje committed:
when deleting cacheable, try to cancel them too

deeje committed:
return pull results in callback

deeje committed:
(fix a typo in ReadMe)

deeje committed:
(more ReadMe edits for clarity)

deeje committed:
(updated authors attributions)

deeje committed:
(update comparison w NSPersistentCloudKitCOntainer)

deeje committed:
implement CacheState.unload

deeje committed:
update example to call removeLocal()

deeje committed:
Merge pull request #28 from deeje/feature/Cacheable

Feature/Cacheable

deeje committed:
fix compiler errors in Example app

deeje committed:
Merge pull request #31 from deeje/fix/example5.0

Update Example app

deeje committed:
update version to 5.0.2

deeje committed:
Merge pull request #32 from deeje/fix/versionInSpec

5.0.2 - Update Version in Podspec

deeje committed:
progress only increases, qos is configurable

deeje committed:
bump podspec version to 5.1.0

deeje committed:
always use persistentHistory, reduce churn on rapid saves

remove old code paths

deeje committed:
need to update removeStatus locally to stay in sync

deeje committed:
simplify the .upload triggering

deeje committed:
increase delay in sync to 5 sec after last save

deeje committed:
cacheable ops can already be in running

deeje committed:
restart all long-lived ops if one errors out

deeje committed:
if CKErrorRetryAfterKey, wait to push more changes

deeje committed:
centralize the pause timer

deeje committed:
share MOC across CoreDataObserver & CacheManager

for processing the push of history and upload of cacheables

deeje committed:
push operations are atomic

deeje committed:
rapid inserts could cause converter to fail

why use yet another background context?!

deeje committed:
(remove testing code)

deeje committed:
more cleanup of ObjectToRecordOperation

deeje committed:
background tasks are only on iOS

deeje committed:
update ReadMe to include links to MediaBook app

deeje committed:
NSPersistentHistoryTracking is now required

deeje committed:
Merge pull request #33 from deeje/feature/5.1

5.1 - improved sync and upload
